summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/FutureQueue.scala

```diff
@@ -0,0 +1,150 @@
+import com.twitter.util.{ Await, Future, Promise }
+import java.util.concurrent.TimeoutException
+import java.util.concurrent.atomic.AtomicInteger
+import org.slf4j.{ Logger, LoggerFactory }
+import scala.util.{ Failure, Success, Try }
+
+private[summingbird] object FutureQueue {
+  /**
+   * Ratio of total number of outstanding futures to the portion that is finished
+   * ,at which finished futures are cleared.
+   * Clearing finished futures costs proportional to total number of outstanding
+   * futures, so we want to make sure we only clear when sufficient portion is
+   * finished.
+   */
+  val OutstandingFuturesDequeueRatio = 2
+
+  /**
+   * Wait for n futures to finish. Doesn't block, the returned future is satisfied
+   * once n futures have finished either successfully or unsuccessfully.
+   * If n is greater than number of futures in queue then we wait on all of them.
+   */
+  def waitN[A](fs: Iterable[Future[A]], n: Int): Future[Unit] = {
+    val waitOnCount = Math.min(fs.size, n)
+    if (waitOnCount <= 0) {
+      Future.Unit
+    } else {
+      val count = new AtomicInteger(waitOnCount)
+      val p = Promise[Unit]()
+      fs.foreach { f =>
+        f.ensure {
+          // Note that since we are only decrementing we can cross 0 only
+          // once (unless we decrement more than 2^32 times).
+          if (count.decrementAndGet() == 0) {
+            p.setValue(())
+          }
+        }
+      }
+      p
+    }
+  }
+}
+
+private[summingbird] class FutureQueue[S, T](
+    maxWaitingFutures: MaxWaitingFutures,
+    maxWaitingTime: MaxFutureWaitTime,
+    maxEmitPerExec: MaxEmitPerExecute) {
+  @transient protected lazy val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  private[executor] lazy val outstandingFutures = Queue.linkedNonBlocking[Future[Unit]]
+  private lazy val numPendingOutstandingFutures = new AtomicInteger(0)
+  private lazy val responses = Queue.linkedNonBlocking[(S, Try[T])]
+
+  def addAll(iter: TraversableOnce[(S, Future[T])]): Unit = {
+    val addedSize = iter.foldLeft(0) {
+      case (size, (state, fut)) =>
+        val responded =
+          fut
+            .onSuccess { t => responses.put((state, Success(t))) }
+            .onFailure { t => responses.put((state, Failure(t))) }
+        // Make sure there are not too many outstanding:
+        if (addOutstandingFuture(responded.unit)) {
+          size + 1
+        } else {
+          size
+        }
+    }
+
+    if (outstandingFutures.size > maxWaitingFutures.get) {
+      /*
+       * This can happen on large key expansion.
+       * May indicate maxWaitingFutures is too low.
+       */
+      logger.debug(
+        "Exceeded maxWaitingFutures({}), put {} futures", maxWaitingFutures.get, addedSize
+      )
+    }
+  }
+
+  def addAllFuture(state: S, iterFut: Future[TraversableOnce[(S, Future[T])]]): Unit =
+    addOutstandingFuture(
```

Comment:
can we use `transform` here as well. Not so crazy about the `.onSuccess` `onFailure` while we are making changes.